### PR TITLE
Unique Time Window Extension Pull Request

### PR DESCRIPTION
**Pull request by yashothara**

### CODE REVIEW

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+     */
+    private String generateKey(StreamEvent event) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (VariableExpressionExecutor executor : variableExpressionExecutors) {
+            stringBuilder.append(event.getAttribute(executor.getPosition()));
+        }
+        return stringBuilder.toString();
+    }
+}
+
```
Comment (hmrajas) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Remove empty lines.

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+     * @param matchingEvent the event to be matched with the events at the processor
+     * @param finder        the execution element responsible for finding the corresponding events that matches
+     *                      the matchingEvent based on pool of events at Processor
+     * @return the matched events
+     */
+    @Override
+    public synchronized StreamEvent find(ComplexEvent matchingEvent, Finder finder) {
+        return finder.find(matchingEvent, expiredEventChunk, streamEventCloner);
+    }
+
```
Comment (hmrajas) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Remove empty lines.

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java~ [hunk 3]
```diff
@@ -0,0 +1,276 @@
+/*
```
Comment (hmrajas) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java~]: Remove the file.

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    protected void init(ExpressionExecutor[] attributeExpressionExecutors, ExecutionPlanContext executionPlanContext) {
+        this.executionPlanContext = executionPlanContext;
+        this.expiredEventChunk = new ComplexEventChunk<StreamEvent>(false);
+        variableExpressionExecutors = new VariableExpressionExecutor[attributeExpressionExecutors.length - 1];
+        if (attributeExpressionExecutors.length == 2) {
+                variableExpressionExecutors[0] = (VariableExpressionExecutor) attributeExpressionExecutors[0];
+            if (attributeExpressionExecutors[1] instanceof ConstantExpressionExecutor) {
+                if (attributeExpressionExecutors[1].getReturnType() == Attribute.Type.INT) {
+                    timeInMilliSeconds = (Integer) ((ConstantExpressionExecutor) attributeExpressionExecutors[1]).getValue();
+
```
Comment (Kanapriya) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Format the code properly

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.wso2.siddhi.core.ExecutionPlanRuntime;
+import org.wso2.siddhi.core.SiddhiManager;
+import org.wso2.siddhi.core.event.Event;
+import org.wso2.siddhi.core.query.output.callback.QueryCallback;
+import org.wso2.siddhi.core.stream.input.InputHandler;
+import org.wso2.siddhi.core.util.EventPrinter;
+
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Please add class level comments

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        removeEventCount = 0;
+        eventArrived = false;
+    }
+
+    /**
+     * Commenting out intermittent failing test case until fix this properly.
+     */
+    @Test
+    public void uniqueTimeWindowTest1() throws InterruptedException {
+
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Remove empty line

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        executionPlanRuntime.addCallback("query1", new QueryCallback() {
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                EventPrinter.print(timeStamp, inEvents, removeEvents);
+                if (inEvents != null) {
+                    inEventCount = inEventCount + inEvents.length;
+                }
+                eventArrived = true;
+            }
+
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Remove unwanted lines , please check these kind of fix throughout the PR's

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Remove unwanted lines

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        inputHandler.send(new Object[]{"IBM", 60.4f, 4});
+        Thread.sleep(1100);
+        inputHandler.send(new Object[]{"RRR", 700f, 3});
+        inputHandler.send(new Object[]{"WSO2", 60.5f, 7});
+        Thread.sleep(1100);
+        Assert.assertEquals(5, inEventCount);
+        Assert.assertTrue(eventArrived);
+        executionPlanRuntime.shutdown();
+    }
+
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Add method level comments. Fix these kind of issues throughout the PR's

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                if (inEvents != null) {
+                    inEventCount = inEventCount + inEvents.length;
+                }
+                eventArrived = true;
+            }
+
+        });
+        InputHandler inputHandler = executionPlanRuntime.getInputHandler("cseEventStream");
+        executionPlanRuntime.start();
+        inputHandler.send(new Object[]{"IBM", 700f, 1});
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Get these ( "IBM", 700f, 1) values from constant file

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        inputHandler.send(new Object[]{"WSO2", 60.5f, 7});
+        Thread.sleep(1100);
+        inputHandler.send(new Object[]{"IBM", 700f, 5});
+        inputHandler.send(new Object[]{"AAA", 60.5f, 6});
+        Thread.sleep(500);
+        Assert.assertEquals(0, inEventCount);
+        Assert.assertTrue(eventArrived);
+        executionPlanRuntime.shutdown();
+    }
+
```
Comment (Kanapriya) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Remove unwanted lines

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java (hunk 2 above)
Comment (yashothara) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: pom.xml [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>org.wso2.extension.siddhi.window.uniquetime</Bundle-SymbolicName>
+                        <Bundle-Name>org.wso2.extension.siddhi.window.uniquetime</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.window.uniquetime,
+                            org.wso2.extension.siddhi.window.uniquetime.*
```
Comment (yashothara) [inline pom.xml]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+/*
+* Sample Query:
+* from inputStream#window.unique:time(attribute1,3 sec)
+* select attribute1, attribute2
+* insert into outputStream;
+*
+* Description:
+* In the example query given, 3 is the duration of the window and attribute1 is the unique attribute.
+* According to the given attribute it will give unique events within given time.
+* */
```
Comment (yashothara) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java (hunk 4 above)
Comment (yashothara) [inline src/main/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowProcessor.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 5 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 6 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 7 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 8 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 9 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 10 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

File: src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java (hunk 11 above)
Comment (yashothara) [inline src/test/java/org/wso2/extension/siddhi/window/uniquetime/UniqueTimeWindowTestCase.java]: Fixed in https://github.com/wso2-extensions/siddhi-window-unique-time/pull/7.

**Pull request merged**